### PR TITLE
Fix curly braces

### LINKED ISSUE #727: 📘 DOC: Opening Curly Brace shows as HTML Encoded Value ("Built-In Components")
image

[Website Reference](https://docs.astro.build/reference/builtin-components)

If you look at the [source file](https://github.com/snowpackjs/astro/blob/main/docs/src/pages/reference/builtin-components.md), then you will notice that it does not look incorrect as shown in the screenshot.  I actually can't figure out why it is happening so this might also be a bug within Astro SSR.

### PR DESCRIPTION
**Pull request by mmarkelov**
## Changes

Should fix #727

Absolutely not sure about this fix.
I just found that if we use `'&#123;'` and try to replace it, we got unexpected result and braces are stayed as symbol. 
So I decided to use some custom constant to replace `{` with it and then we can replace constant to `{` during transformation.   

- What does this change?
- Be short and concise. Bullet points can help!
- Before/after screenshots can be helpful as well.

### CODE REVIEW

**Comment by matthewp** [comment]
Can you add a test?

File: packages/astro/src/compiler/codegen/index.ts [hunk 1]
```diff
@@ -225,7 +224,7 @@ function getComponentWrapper(_name: string, hydration: HydrationAttributes, { ur
       }
     };
 
-    let metadata: string = '';
+    let metadata = '';
```
Comment (jasikpark) [inline packages/astro/src/compiler/codegen/index.ts:227]: why remove this type? is this variable not a string?

Comment (matthewp) [reply packages/astro/src/compiler/codegen/index.ts:227]: The value is a string so the type isn't necessary

**Comment by eyelidlessness** [comment]
If I had to educated guess, this is caused by a workaround of a deeper problem wherein static string content is being treated as expressions. This is something I raised while prototyping the Solid renderer and discovered a variety of XSS/HTML injection scenarios related to treating renderer SSR output as HTML.

I suggested then, and I’ll reiterate now, that there should be a way to differentiate strings from template expressions so this doesn’t keep happening. These escapes should be removed entirely, and Astro (and Astro renderers) should have a clear way to distinguish actual strings from build-time strings which represent HTML or Astro children and template expressions. Otherwise this will be a never-ending whack-a-mole.

**Comment by mmarkelov** [comment]
@eyelidlessness i absolutely agree with you. I'll just try to fix this, but not sure if i can, cause i don't have any experience before.

**Comment by mmarkelov** [comment]
> Can you add a test?

@matthewp  added. Also tried to fix it in nice way, but it's a complicated :(
So idk what should we do. I tried to look through all possible cases with this changes - and it's ok. But I just don't want to break things with this one.

**Comment by natemoo-re** [comment]
@mmarkelov thank you for the PR! You are deep in the weeds on this one. 😅 

@eyelidlessness is correct, this is absolutely a hack that is symptomatic of a larger problem. This is the biggest motivation for rewriting our forthcoming parser/compiler from scratch.

Until that lands, I'm not sure the best way to proceed. Do we merge this, fixing the original hack with a new hack? I'd like to at least investigate how much effort it would take to partially fix the underlying problem before moving forward with this.

**Comment by jonathantneal** [comment]
> Do we merge this, fixing the original hack with a new hack? I'd like to at least investigate how much effort it would take to partially fix the underlying problem before moving forward with this.

Considering this has impacted a few folks in Discord, and considering the impact it has on Astro documentation, I would suggest improving the _existing_ hack, with as little impact as possible. This will offer an immediate reward of less breakage, and a permanent reward of improved test coverage. @mmarkelov, @natemoo-re, for your consideration — https://github.com/snowpackjs/astro/pull/1094 (preserving your work and credit)

**Review by matthewp — APPROVED** [review]
lgtm

**Comment by matthewp** [comment]
We should merge, the fix looks good to me.

**Comment by mmarkelov** [comment]
> We should merge, the fix looks good to me.

I think that it will be better to merge #1094

**Comment by natemoo-re** [comment]
Closed in https://github.com/snowpackjs/astro/pull/1094

**Pull request closed without merging**